### ISSUE TITLE
apex mixed precision performance comparisons

### ISSUE THREAD

**Issue by tmac1997** [NVIDIA/apex#478]
I've done a lot of performance comparisons between apex (python-build version) and myself-implemented mixed precision code on ResNet, PreActResNet, ResNeXt and DenseNet with CIFAR-10. I found a few things really puzzling me:
- It seems myself-implemented mixed percision code actually works better in speed up training performance than apex. I simply follow the guide in quick start and added a few lines into my code. Did I miss something?
- For densenet, it seems `opt_level=O1` performance worse as it took more time to complete training. Why this would happend?

Here is my comparison results. All the experiments are conducted in a single V100 and the package versions are the same for both of them.

For apex:

<table>
  <tr>
    <th rowspan="2">Model</th>
    <th colspan="3">Opt Level (O0 / O1 / O2 / O3)</th>
  </tr>
  <tr>
    <td>Acc</td>
    <td>Time</td>
    <td>GPU</td>
  </tr>
  <tr>
    <td>ResNet152</td>
    <td>93.42 / 93.55 / 93,50 / 93.21</td>
    <td>6h54min / 5h26min / 4h31min / 3h52min</td>
    <td>12855MiB / 7499MiB / 7425MiB / 6915MiB</td>
  </tr>
  <tr>
    <td>PreActResNet152</td>
    <td>93.27 / 93.39 / 93.34 / 93.65</td>
    <td>6h44min / 5h26min / 4h24min / 3h47min</td>
    <td>12593MiB / 7323MiB / 7079MiB / 6783MiB</td>
  </tr>
  <tr>
    <td>ResNext29_32x64d</td>
    <td>94.04 / 93.84 / 94.01 / 93.63</td>
    <td>2h46min / 2h29min / 2h15min / 2h04min</td>
    <td>6348MiB / 3379MiB / 3673MiB / 3647MiB</td>
  </tr>
  <tr>
    <td>DenseNet121</td>
    <td>93.38 / 93.79 / 93.75 / 93.54</td>
    <td>3h13min / 3h57min / 2h56min / 2h16min</td>
    <td>8823MiB / 5133MiB / 5129MiB / 5085MiB</td>
  </tr>
</table>

For implemented mixed precision code:

<table>
  <tr>
    <th rowspan="2">Model</th>
    <th colspan="3">Standard</th>
    <th colspan="3">Mixed Precision(o scale)</th>
    <th colspan="3">Mixed Precision(w scale)</th>
  </tr>
  <tr>
    <td>Acc</td>
    <td>Time</td>
    <td>GPU</td>
    <td>Acc</td>
    <td>Time</td>
    <td>GPU</td>
    <td>Acc</td>
    <td>Time</td>
    <td>GPU</td>
  </tr>
  <tr>
    <td>ResNet152</td>
    <td>93.42</td>
    <td>6h54min</td>
    <td>12855MiB</td>
    <td>93.44</td>
    <td>3h49min</td>
    <td>7609MiB</td>
    <td>93.40</td>
    <td>3h37min</td>
    <td>7601MiB</td>
  </tr>
  <tr>
    <td>PreActResNet152</td>
    <td>93.27</td>
    <td>6h44min</td>
    <td>12593MiB</td>
    <td>93.31</td>
    <td>3h42min</td>
    <td>7441MiB</td>
    <td>93.42</td>
    <td>3h35min</td>
    <td>7575MiB</td>
  </tr>
  <tr>
    <td>ResNext29_32x64d</td>
    <td>94.04</td>
    <td>2h46min</td>
    <td>6438MiB</td>
    <td>93.84</td>
    <td>1h59min</td>
    <td>3781MiB</td>
    <td>93.70</td>
    <td>1h59min</td>
    <td>3783MiB</td>
  </tr>
  <tr>
    <td>DenseNet121</td>
    <td>93.38</td>
    <td>3h13min</td>
    <td>8823MiB</td>
    <td>93.78</td>
    <td>2h18min</td>
    <td>5235MiB</td>
    <td>93.78</td>
    <td>2h19min</td>
    <td>5241MiB</td>
  </tr>
</table>

**Comment by ptrblck** [comment]
Hi @tmac1997,

thanks for sharing your profiling.
Could you elaborate what you mean by the first point? What does "self-implemented" refer to?
In your results I can only find `amp`'s `opt_levels`.

**Comment by tmac1997** [comment]
HI @ptrblck,
here I mean I write mixed precision training code myself - change the model to FP16, keep the FP32 master copy and update in FP32. The first table shows the results of NVIDIA apex and the second table is about my own implementation.
What surprises me is that my own implementation is a little bit faster than NVIDIA apex. I'm wondering if I use NVIDIA apex in the proper way.

**Comment by ptrblck** [comment]
Thanks for the information!
Did you also keep the batchnorm layers in FP32 and use loss scaling ([lines of code](https://github.com/NVIDIA/apex/blob/325f5a0bec542701edba1628ad34f3b2ea47c556/apex/amp/frontend.py#L139))?

It would be interesting to see, where the overhead comes from, if you've implemented exactly the same mixed precision recipe.

**Comment by lynnliu030** [comment]
@syorami May I ask what batch size you use for resNet152 and how many epochs you run to produce this comparison result?
